worker/src/d1.rs

```diff
@@ -0,0 +1,216 @@
+        let array = Uint8Array::new(&array_buffer);
+        let mut vec = Vec::with_capacity(array.length() as usize);
+        array.copy_to(&mut vec);
+        Ok(vec)
+    }
+
+    /// Batch execute one or more statements against the database.
+    ///
+    /// Returns the results in the same order as the provided statements.
+    pub async fn batch(&self, statements: Vec<D1PreparedStatement>) -> Result<Vec<D1Result>> {
+        let statements = statements.into_iter().map(|s| s.0).collect::<Array>();
+        let results = JsFuture::from(self.0.batch(statements)).await?;
+        let results = results.dyn_into::<Array>()?;
+        let mut vec = Vec::with_capacity(results.length() as usize);
+        for result in results.iter() {
+            let result = result.dyn_into::<D1ResultSys>()?;
+            vec.push(D1Result(result));
+        }
+        Ok(vec)
+    }
+
+    /// Execute one or more queries directly against the database.
+    ///
+    /// The input can be one or multiple queries separated by `\n`.
+    ///
+    /// # Considerations
+    ///
+    /// This method can have poorer performance (prepared statements can be reused
+    /// in some cases) and, more importantly, is less safe. Only use this
+    /// method for maintenance and one-shot tasks (example: migration jobs).
+    ///
+    /// If an error occurs, an exception is thrown with the query and error
+    /// messages, execution stops and further statements are not executed.
+    pub async fn exec(&self, query: &str) -> Result<D1ExecResult> {
+        let result = JsFuture::from(self.0.exec(query)).await?;
+        Ok(result.into())
+    }
+}
+
+impl EnvBinding for D1Database {
+    const TYPE_NAME: &'static str = "D1Database";
+}
+
+impl JsCast for D1Database {
+    fn instanceof(val: &JsValue) -> bool {
+        val.is_instance_of::<D1DatabaseSys>()
+    }
+
+    fn unchecked_from_js(val: JsValue) -> Self {
+        Self(val.into())
+    }
+
+    fn unchecked_from_js_ref(val: &JsValue) -> &Self {
+        unsafe { &*(val as *const JsValue as *const Self) }
+    }
+}
+
+impl From<D1Database> for JsValue {
+    fn from(database: D1Database) -> Self {
+        JsValue::from(database.0)
+    }
+}
+
+impl AsRef<JsValue> for D1Database {
+    fn as_ref(&self) -> &JsValue {
+        &self.0
+    }
+}
+
+impl From<D1DatabaseSys> for D1Database {
+    fn from(inner: D1DatabaseSys) -> Self {
+        Self(inner)
+    }
+}
+
+// A D1 prepared query statement.
+pub struct D1PreparedStatement(D1PreparedStatementSys);
+
+impl D1PreparedStatement {
+    /// Bind one or more parameters to the statement. Returns a new statement object.
+    ///
+    /// D1 follows the SQLite convention for prepared statements parameter binding.
+    ///
+    /// # Considerations
+    ///
+    /// Supports Ordered (?NNNN) and Anonymous (?) parameters - named parameters are currently not supported.
+    ///
+    pub fn bind<T>(&self, values: &[&T]) -> Result<Self>
```

Comment:
`bind` only allows for a single type in the array. This prevents you from doing something like

```rust
db.prepare("INSERT INTO users (email, email_verified) VALUES (?1, ?2)").bind(&[&"contact@example.com", &true])
```

Using a macro that converts the values into `JsValue` before binding would circumvent this issue. Following what `sqlx` does would be a good solution (https://docs.rs/sqlx/latest/sqlx/macro.query.html#query-arguments).

```rust
worker::d1::query!("INSERT INTO users (email, email_verified) VALUES (?1, ?2)", "contact@example.com", true)
```

Reply:
This is a good point, and something I wanted to do initially. I'll take a look and see if I can implement a macro that returns a pre-bound statement.

Reply:
After playing around a bit I think the following macro definition would work.

```rs
macro_rules! query {
    ($db:expr, $query:expr) => {
        $crate::d1::Database::prepare($db, $query)
    };
    ($db:expr, $query:expr, $($args:expr),* $(,)?) => {{
        || -> ::std::result::Result<$crate::d1::PreparedStatement, ::worker::Error> {
            let prepared = $crate::d1::Database::prepare($db, $query);

            // D1 does not accept undefined as NULL
            let serializer = ::serde_wasm_bindgen::Serializer::new().serialize_missing_as_null(true);
            let bindings = &[$(
                ::serde::ser::Serialize::serialize(&$args, &serializer)
                    .map_err(|e| ::worker::Error::Internal(e.into()))?
            ),*];

            $crate::d1::PreparedStatement::bind(&prepared, bindings)
        }()
    }};
}

pub(crate) use query;
```

Reply:
I'm going to take a look now and see if I can get this implemented. 👍

Reply:
@SebastiaanYN I've just pushed an implementation of the `query!` macro. I had to do a little bit of tinkering to get the crate working in an example, I've included a re-export of `serde-wasm-bindgen` from `worker` so that consumer modules don't need to import it themselves.